compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -655,6 +700,23 @@ macro_rules! define_queries {
             }
         }
 
+        // These arrays are used for iteration and can't be indexed by `DepKind`.
+
+        const TRY_COLLECT_ACTIVE_JOBS: &[for<'tcx> fn(TyCtxt<'tcx>, &mut QueryMap<DepKind>)] =
+            &[$(query_impl::$name::try_collect_active_jobs),*];
+
+        const ALLOC_SELF_PROFILE_QUERY_STRINGS: &[
+            for<'tcx> fn(TyCtxt<'tcx>, &mut QueryKeyStringCache)
+        ] = &[$(query_impl::$name::alloc_self_profile_query_strings),*];
+
+        const ENCODE_QUERY_RESULTS: &[
+            Option<for<'tcx> fn(
+                TyCtxt<'tcx>,
+                &mut CacheEncoder<'_, 'tcx>,
+                &mut EncodedDepNodeIndex)
+            >
+        ] = &[$(expand_if_cached!([$($modifiers)*], query_impl::$name::encode_query_results)),*];
```

Comment:
Can this use `item_if_cached`? Like this:
```suggestion
        ] = &[$( item_if_cached!([$($modifiers)*], query_impl::$name::encode_query_results, ))*];
```

This will allow to remove the option.

Reply:
The macro has to expand to a single expression, so it's not quite that simple.